fix: Blank selected department when selecting out of range

rique223 opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  feat: Adding a new feature
  refactor: A code change that doesn't change behavior (it doesn't add anything and doesn't fix anything)
  fix: For bug fixes that affect the end-user
  chore: For small tasks
  docs: For documentation
  ci: For updating CI configuration
  test: For adding tests
  i18n: For updating any translations
  regression: Issues created/reported/fixed during the development phase. kind of problem that never existed in production and that we don't need to list in a changelog for the end user
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description won't be displayed to our end users in the release notes, so feel free to add as much technical context as needed.
  If the changes introduced in this pull request must be presented in the release notes, make sure to add a changeset file. Check our guidelines for adding a changeset to your pull request: https://developer.rocket.chat/contribute-to-rocket.chat/modes-of-contribution/participate-in-rocket.chat-development/development-workflow#4.-adding-changeset-to-your-pull-request 
-->
The Omnichannel current chats departments filter uses "on scroll" pagination (it fetches a list of the 25 first departments and loads more items every 25 items that the user scrolls down). If a user selects a department that isn’t within the first 25 departments listed, the selection will appear blank after refreshing the page. This happens because the `PaginatedSelectFiltered` component cannot find the selected department in the paginated options list due to it not being part of the initially fetched list of departments. This PR adds a "normalization" function to the listing of departments that will add the missing department to the initial list so the component can correctly show the previously selected department.

Before:
![oc-departments-before](https://github.com/user-attachments/assets/896cad96-eceb-4057-af1e-ab0f9e17806a)

After:
![oc-departments-after](https://github.com/user-attachments/assets/96306d92-6502-419e-a9e3-493f5b966271)

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->
Jira task: [SUP-675](https://rocketchat.atlassian.net/browse/SUP-675)

[SUP-675]: https://rocketchat.atlassian.net/browse/SUP-675?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

KevLehman reviewed (changes requested):
Can we add a couple of e2e tests?

rique223 commented:
> Can we add a couple of e2e tests?

Hey @KevLehman, talking to the frontend folks we concluded that it would be better to test this specific behavior through unit tests of the `useDepartmentLists` hook so I've added the new unit tests and also added a new test to verify the localStorage usage of the departments filter.

KevLehman reviewed (approved):
I'll remove the omni review 🤗

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/hooks/useDepartmentsList.spec.ts:
```diff
@@ -0,0 +1,107 @@
+	};
+});
+
+it('should not add selected department if it is already in the departments list on first fetch', async () => {
+	const selectedDepartmentMappedToOption = {
+		_id: '5',
+		label: 'test_department_5',
+		value: '5',
+	};
+
+	const { result } = renderHook(
+		() =>
+			useDepartmentsList({
+				filter: '',
+				onlyMyDepartments: true,
+				haveAll: true,
+				showArchived: true,
+				selectedDepartment: '5',
+			}),
+		{
+			legacyRoot: true,
+			wrapper: mockAppRoot()
+				.withEndpoint('GET', '/v1/livechat/department', () => ({
+					count: 25,
+					offset: 0,
+					total: 25,
+					departments: initialDepartmentsListMock,
+				}))
+				.build(),
+		},
+	);
+
+	await waitFor(() => expect(result.current.itemsList.items).toContainEqual(selectedDepartmentMappedToOption));
+	// The expected length is 26 because the hook will add the 'All' item on run time
+	await waitFor(() => expect(result.current.itemsList.items.length).toBe(26));
+});
```
I believe this test case could be improved by mocking the endpoint `/v1/livechat/department/:_id` and making sure it is not being called. This can be achieved with a combination of `withEndpoint` and `jest.fn()`. If you need help implementing, give me a heads up.

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/normalizeDepartments.ts:
```diff
@@ -0,0 +1,20 @@
+	departments: DepartmentListItem[],
+	selectedDepartment: string,
+	getDepartment: EndpointFunction<'GET', '/v1/livechat/department/:_id'>,
+) => {
```
I'd recommend adding a return type in this case just to ensure the item is being formatted correctly.

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/normalizeDepartments.ts:
```diff
@@ -0,0 +1,20 @@
+	selectedDepartment: string,
+	getDepartment: EndpointFunction<'GET', '/v1/livechat/department/:_id'>,
+) => {
+	const isSelectedDepartmentAlreadyOnList = departments.find((department) => department._id === selectedDepartment);
```
Two things. 
1. Using `.some()` is more appropriate for this scenario
2. `isSelectedDepartmentAlreadyOnList` could be a function to prevent searching the array unnecessarily.

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/normalizeDepartments.ts:
Create a `/utils` folder for this function

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/hooks/useDepartmentsList.ts:
```diff
@@ -14,9 +15,10 @@ type DepartmentsListOptions = {
 };
 
-type DepartmentListItem = {
+export type DepartmentListItem = {
```
I recommend moving this to a separate module to avoid the cyclic dependency with `normalizeDepartments`

aleksandernsilva commented on apps/meteor/client/components/Omnichannel/utils/normalizeDepartments.ts:
```diff
@@ -0,0 +1,20 @@
+		return departments;
+	}
+
+	const { department: missingDepartment } = await getDepartment({});
+
+	return missingDepartment
+		? [...departments, { _id: missingDepartment._id, label: missingDepartment.name, value: missingDepartment._id }]
+		: departments;
+};
```
Just to be on the safe side, might be a good idea to add a try catch for this fetch and guarantee that departments is returned even if something goes wrong.
